index.js

```diff
@@ -0,0 +1,214 @@
+    typeof body.scores === 'object' && body.scores !== null,
+    400,
+    '.scores should be an object'
+  )
+  httpAssert(
+    Object.keys(body.scores).every(ethers.isAddress),
+    400,
+    'All .scores keys should be 0x addresses'
+  )
+  httpAssert(
+    Object.values(body.scores).every(score => {
+      try {
+        return BigInt(score) > 0n
+      } catch {
+        return false
+      }
+    }),
+    400,
+    'All .scores values should be positive numbers encoded as string'
+  )
+
+  validateSignature(body.signature, body.scores, signerAddresses)
+
+  const timestamp = new Date()
+  const tx = redis.multi()
+  for (const [address, score] of Object.entries(body.scores)) {
+    const scheduledRewards = (BigInt(score) * roundReward) / maxScore
+    tx.hincrby('rewards', address, scheduledRewards)
+    tx.rpush(
+      'log',
+      JSON.stringify({
+        timestamp,
+        address,
+        score,
+        scheduledRewards: String(scheduledRewards)
+      })
+    )
+  }
+  const updated = await tx.exec()
+
+  json(
+    res,
+    Object.fromEntries(
+      Object
+        .keys(body.scores)
+        .map((address, i) => [address, String(updated[i * 2][1])])
+    )
+  )
+}
+
+async function handlePaidScheduledRewards (req, res, redis, signerAddresses) {
+  const body = JSON.parse(await getRawBody(req, { limit: '1mb' }))
+
+  httpAssert(
+    typeof body === 'object' && body !== null,
+    400,
+    'Request body should be an object'
+  )
+  httpAssert(
+    typeof body.rewards === 'object' && body.rewards !== null,
+    400,
+    '.rewards should be an object'
+  )
+  httpAssert(
+    Object.keys(body.rewards).every(ethers.isAddress),
+    400,
+    'All .rewards keys should be 0x addresses'
+  )
+  httpAssert(
+    Object.values(body.rewards).every(score => {
+      try {
+        return BigInt(score) > 0n
+      } catch {
+        return false
+      }
+    }),
+    400,
+    'All .rewards values should be positive numbers encoded as string'
+  )
+
+  validateSignature(body.signature, body.rewards, signerAddresses)
+
+  const timestamp = new Date()
+  const tx = redis.multi()
+  for (const [address, amount] of Object.entries(body.rewards)) {
+    tx.hincrby('rewards', address, BigInt(amount) * -1n)
+    tx.rpush(
+      'log',
+      JSON.stringify({
+        timestamp,
+        address,
+        scheduledRewards: String(BigInt(amount) * -1n)
+      })
+    )
+  }
+  const updated = await tx.exec()
+
+  json(
+    res,
+    Object.fromEntries(
+      Object
+        .keys(body.rewards)
+        .map((address, i) => [address, String(updated[i * 2][1])])
+    )
+  )
+}
+
+async function handleGetAllScheduledRewards (res, redis) {
+  json(res, await redis.hgetall('rewards'))
+}
+
+async function handleGetSingleScheduledRewards (req, res, redis) {
+  json(res, await redis.hget('rewards', req.url.split('/').pop()))
+}
+
+async function handleGetLog (res, redis) {
+  json(res, (await redis.lrange('log', 0, -1)).map(JSON.parse))
```

Comment:
It's not very efficient to parse all log entries to objects and then serialize them back to JSON, but I guess we don't care about that now.

Options we could consider:
- Return NDJSON instead - this will make it easy to process the response in a streaming way. We could produce the response incrementally too, if ioredis provided a streaming variant of `lrange`.
- Build the JSON array via string manipulation:
   ```js
   const body = '[' + (await redis.lrange('log', 0, -1)).join(',') + ']'
   // set content type, write the body
   ```

No changes in your PR are needed; I'm just sharing my ideas.

Reply:
Thank you so much for measuring this!

if 10 days of data takes 2.2s, and we want to keep 30 days, then I agree this is something to optimize, very much

Reply:
I'll try string concatenation first, as when we produce JSON and not NDJSON it's easier to consume for humans (not tools). We can freely change this API later, as it only serves for debugging